File: 5.toutiao_news/toutiao_front_end/src/api/favorite.ts
/**
 * 收藏相关 API 接口封装
 */
import request from '@/utils/request'

// ============ 请求参数类型 ============
/** 收藏列表分页参数 */
interface FavoriteListParams {
  /** 当前页码 */
  page?: number
  /** 每页条数 */
  pageSize?: number
}

/**
 * 获取收藏列表（分页）
 * @param params - 分页参数
 * @returns 收藏列表的接口响应
 */
export function getFavoriteList(params: FavoriteListParams): Promise<ApiResponse> {
  return request({
    url: '/favorite/list',
    method: 'get',
    params
  })
}

/**
 * 添加收藏
 * @param newsId - 要收藏的新闻ID
 * @returns 接口响应
 */
export function addFavorite(newsId: number): Promise<ApiResponse> {
  return request({
    url: '/favorite/add',
    method: 'post',
    data: { newsId }
  })
}

/**
 * 取消收藏
 * @param newsId - 要取消收藏的新闻ID
 * @returns 接口响应
 */
export function removeFavorite(newsId: number): Promise<ApiResponse> {
  return request({
    url: '/favorite/remove',
    method: 'delete',
    params: { newsId }
  })
}

/**
 * 清空所有收藏
 * @returns 接口响应，data.count 表示本次清空的收藏条数
 */
export function clearFavorites(): Promise<ApiResponse<{ count: number }>> {
  return request({
    url: '/favorite/clear_all_favorite',
    method: 'delete'
  })
}

/**
 * 检查某条新闻是否已收藏（详情页进入时判断收藏状态）
 * @param newsId - 新闻ID
 * @returns 接口响应，data.isFavorited 表示是否已收藏
 */
export function checkFavorite(newsId: number): Promise<ApiResponse<{ isFavorited: boolean }>> {
  return request({
    url: '/favorite/check',
    method: 'get',
    params: { newsId }
  })
}

File: 5.toutiao_news/toutiao_back_end/routers/favorite.py
from fastapi import HTTPException

from fastapi import APIRouter
from fastapi import Depends
from config.db_conf import get_db
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import Session

from crud import  favorite
from models.users import User
from schemas.favorite import FavoriteCheckResponse, FavoriteAddResponse, FavoriteListResponse
from utils.auth import get_current_user
from fastapi import Query, Body

from utils.response import success_response

router = APIRouter(
    prefix="/api/favorite",
    tags=["favorite"],
    # dependencies=[Depends(get_current_user)],
)

# 检查新闻收藏状态
@router.get('/check')
async def check_favorite(
        db: AsyncSession = Depends(get_db),
        user: User = Depends(get_current_user),
        new_id: int = Query(..., alias="newsId", description="新闻ID"),
):
    checked = await favorite.is_new_favorite(db, user, new_id)

    return success_response(message="未收藏", data=FavoriteCheckResponse(isFavorite=checked))


# 收藏新闻
@router.post('/add')
async def add_favorite(
        data: FavoriteAddResponse,
        db: AsyncSession = Depends(get_db),
        user: User = Depends(get_current_user),
):
    result = await favorite.add_new_favorite(db, user, data.news_id)
    return success_response(message="收藏成功", data=result)

# 取消收藏
@router.delete('/remove')
async def remove_favorite(
        db: AsyncSession = Depends(get_db),
        user: User = Depends(get_current_user),
        news_id: int = Query(..., alias="newsId", description="新闻ID"),
):
    result = await favorite.remove_new_favorite(db, user, news_id)
    if not result:
        raise HTTPException(status_code=404, detail="收藏记录不存在")
    return success_response(message="删除收藏成功")

#  查询收藏列表
@router.get('/list')
async def get_favorite_list(
        db: AsyncSession = Depends(get_db),
        user: User = Depends(get_current_user),
        page_size: int = Query(10, alias="pageSize", description="每页数量"),
        page: int = Query(1, alias="page", description="页码"),
):
    rows, total = await favorite.get_favorite_list(db, user.id, page, page_size)
    # [expression for item in iterable]
    favorite_list = [{
        **news.__dict__,
        "favorite_time": favorite_time,
        "favorite_id": favorite_id
    } for news, favorite_time, favorite_id in rows]
    has_more = total > page_size * page
    data = FavoriteListResponse(list=favorite_list, total=total, hasMore=has_more)
    return success_response(message="获取收藏列表成功", data=data)


#  取消收藏列表
@router.delete('/clear_all_favorite')
async def clear_all_favorite(
        db: AsyncSession = Depends(get_db),
        user: User = Depends(get_current_user),
):
    count = await favorite.remove_all_favorite(db, user.id)
    return success_response(message=f"清空了{count}条数据", data=count)